src/librustc_driver/driver.rs

```diff
@@ -150,16 +150,23 @@ pub fn compile_input(sess: Session,
 
         (outputs, trans, tcx.sess)
     };
-    phase_5_run_llvm_passes(&sess, &trans, &outputs);
-
-    controller_entry_point!(after_llvm,
+    if !sess.targeting_pnacl() {
+        phase_5_run_llvm_passes(&sess, &mut trans, &outputs);
+        controller_entry_point!(after_llvm,
                             sess,
                             CompileState::state_after_llvm(input,
                                                            &sess,
                                                            outdir,
                                                            &trans));
 
-    phase_6_link_output(&sess, &trans, &outputs);
+        phase_6_link_output(&sess, &trans, &outputs);
+    } else {
+        let cid = trans.link.crate_name.clone();
+        link::link_outputs_for_pnacl(&sess,
+                                     &mut trans,
+                                     &outputs,
+                                     &cid[..]);
+    }
```

Comment:
More pnacl specific stuff in the front end is going to cause a lot of noise for other people. Hopefully this can be rationalized into something non-platform-specific, like asking the target spec if it requires bitcode translation, and if not calling a different backend pass. We could also fold phase 5 and phase 6 so we don't have to have some pipelines going all the way through phase 6 and some having a special 'extra' phase (7?).